### PR TITLE
doc: initial design documentation

### PR DESCRIPTION
**Pull request by asmie**
The initial part of the design documentation.

### CODE REVIEW

File: third_party/move/tools/move-mutator/doc/design.md [hunk 1]
```diff
@@ -0,0 +1,252 @@
+# Move mutator tool design
+
+## Overview
+
+The Move mutator tool is a tool that mutates Move source code. It is used to help test the robustness of Move analyzers and verifiers by generating different code versions (mutants).
```
Comment (Rqnsom) [inline third_party/move/tools/move-mutator/doc/design.md]:
```suggestion
The Move mutator is a tool that mutates Move source code. It is used to help test the robustness of Move analyzers and verifiers by generating different code versions (mutants).
```

File: third_party/move/tools/move-mutator/doc/design.md [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+## Prototype
+This chapter will be deleted upon implementing the Move mutator tool.
+
+The Prototype implements only binary operator replacement (arithmetic, bitwise and shifts) and unary operator replacement. Command line options mainly handle source file paths (single or package).
+
+Mutants are currently stored in the `mutants_output` directory, and there is no way to change it. Mutants are just saved in a flat structure (no subdirectories). Each mutant filename consists of the original source filename with the next index appended.
+
+The report is generated on screen and lists all generated mutants (which mutation operator was used, location in the original source file, and mutant output). It's not possible to create the report in a file.
+
+Not supported: configuration file, generating reports, nicely formatted output, mutant files output structure, other mutation operators than binary and unary operator replacement.
```
Comment (Rqnsom) [inline third_party/move/tools/move-mutator/doc/design.md]: nit: Generally, we could use bullet points for such lists to make them more readable

File: third_party/move/tools/move-mutator/doc/design.md [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+
+In this project, the presentation layer execution modules (displaying and fetching data from/to the user) are moved to the `aptos` and `move` command line interfaces.
+
+Mutator itself provides a CLI module, which is integrated into the existing Aptos repository - it handles options that the mutator tool can use.
+
+There are two types of output that the Move mutator tool generates:
+- the actual mutants (source code files)
+- the report (JSON and text file)
+
+The actual mutants are stored in the output directory (default: `mutants_output/mutants`). The directory structure is the same as in the original source code. The mutant filename consists of the original source filename with mutation information (operator) and the next index appended.
```
Comment (Rqnsom) [inline third_party/move/tools/move-mutator/doc/design.md]:
```suggestion
The actual mutants are stored in the output directory (default: `mutants_output/mutants`). The directory structure is the same as in the original source code. The mutant filename consists of the original source filename with mutation information (operator) and the mutant index appended.
```
Unrelated comment:You can insert these two somewhere in the text (e.g. original source code in`module.move`) and (e.g. mutatnt file `module_1.move`).

File: third_party/move/tools/move-mutator/doc/design.md [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+1. Takes arguments both for the Move Prover tool and for the Move mutator tool.
+2. Run the Move mutator tool to generate mutants.
+3. Run the Move Prover tool, passing the generated mutants one by one.
+4. Collect the results and generate a report.
+
+The report contains information about the mutants killed for each tested source file. All the intermediate results are saved in the `mutants_output` directory.
+
+## Mutation operators
+
+Mutation operators can be divided into two groups. The first group is composed of mutation operators that, once replaced, do not change the length of the original program. That operator can be mixed inside on file to achieve more complex mutations. The second group is composed of mutation operators that, once replaced, change the length of the original program. That operator cannot be mixed inside one file, so each mutant has only one mutation of that kind. It should be possible to combine many non-changing file length mutation operators with one changing len mutation operator, but the latter needs to be applied as the last one.
```
Comment (Rqnsom) [inline third_party/move/tools/move-mutator/doc/design.md]: Let's mention explicitly that this is related to running the mutator with such a _generate double mutants option_.

**Review by Rqnsom — APPROVED** [review]
LGTM,
with a few comments.
the parts where code logic is explained could be replaced with some image maybe...

File: third_party/move/tools/move-mutator/doc/design.md (hunk 1 above)
Comment (asmie) [reply third_party/move/tools/move-mutator/doc/design.md]: Done.

File: third_party/move/tools/move-mutator/doc/design.md (hunk 2 above)
Comment (asmie) [reply third_party/move/tools/move-mutator/doc/design.md]: Done.

File: third_party/move/tools/move-mutator/doc/design.md (hunk 4 above)
Comment (asmie) [reply third_party/move/tools/move-mutator/doc/design.md]: Done.

File: third_party/move/tools/move-mutator/doc/design.md (hunk 3 above)
Comment (asmie) [reply third_party/move/tools/move-mutator/doc/design.md]: First part - done.
I'm not feeling right to put names right now - I'm into changing them :smile:

**Review by Rqnsom — APPROVED** [review]

**Pull request merged**